doc/man1/locale.1
.\" $OpenBSD: locale.1,v 1.11 2023/03/05 18:55:34 ajacoutot Exp $
.\"
.\" Copyright 2016, 2020 Ingo Schwarze <schwarze@openbsd.org>
.\" Copyright 2013 Stefan Sperling <stsp@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: March 5 2023 $
.Dt LOCALE 1
.Os
.Sh NAME
.Nm locale
.Nd character encoding and localization conventions
.Sh SYNOPSIS
.Nm locale
.Op Fl a | Fl m | Cm charmap
.Sh DESCRIPTION
If the
.Nm
utility is invoked without any arguments, the current locale
configuration is shown.
Values for categories that are not set in the environment or that are
overridden by
.Ev LC_ALL
are displayed between double quotes.
.Pp
The options are as follows:
.Bl -tag -width charmap
.It Fl a
Display a list of supported locales.
.It Fl m
Display a list of supported character encodings.
On
.Ox ,
this always returns UTF-8 only.
.It Cm charmap
Display the currently selected character encoding.
On
.Ox ,
this returns either US-ASCII or UTF-8.
.El
.Pp
A locale is a set of environment variables telling programs which
character encoding, language and cultural conventions the user
prefers.
Programs in the
.Ox
base system ignore the locale except for the character encoding,
and it is not recommended to use any of these variables except that
the following non-default setting is supported as an option:
.Pp
.Dl export LC_CTYPE=en_US.UTF-8
.Pp
Programs installed from
.Xr packages 7
may or may not change behavior according to the locale.
Many programs use the X/Open System Interfaces naming scheme
for the contents of the variables listed below, which is
.Sm off
.Ar language
.Op _ Ar TERRITORY
.Op \&. Ar encoding
.Op @ Ar modifier
.Sm on
.Pp
The behavior of some library functions may also depend on the locale,
and it does on most other operating systems.
The
.Ox
C library tends to avoid locale-dependent behavior except with
respect to character encoding.
See the manual pages of individual functions for details.
.Pp
The character encoding locale
.Ev LC_CTYPE
instructs programs which character encoding to assume for text input
and to use for text output.
A character encoding maps each character of a given character set
to a byte sequence suitable for storing or transmitting the character.
.Pp
The
.Ox
base system supports two locales: the default of
.Li LC_CTYPE=C
selects the US-ASCII character set and encoding, treating the bytes
0x80 to 0xff as non-printable characters of application-specific
meaning.
.Li LC_CTYPE=POSIX
is an alias for
.Li LC_CTYPE=C .
The alternative of
.Li LC_CTYPE=en_US.UTF-8
selects the UTF-8 encoding of the Unicode character set, which is
supported by many parts of the system, but not yet fully supported
by all parts.
.Pp
If the value of
.Ev LC_CTYPE
ends in
.Ql .UTF-8 ,
programs in the
.Ox
base system ignore the beginning of it, treating for example zh_CN.UTF-8
exactly like en_US.UTF-8.
Programs from
.Xr packages 7
may however make a difference.
If the value of
.Ev LC_CTYPE
is unsupported, programs and libraries in the
.Ox
base systems fall back to
.Li LC_CTYPE=C .
.Pp
Some programs, for example
.Xr write 1 ,
deliberately ignore the locale and always use US-ASCII only.
See the manual pages of individual programs for details.
.Sh ENVIRONMENT
The locale configuration consists of the following environment variables:
.Bl -tag -width LC_MONETARYX
.It Ev LC_ALL
Overrides all other
.Ev LC_*
variables below.
.It Ev LC_COLLATE
Intended to affect collation order.
It may for example affect alphabetic sorting, regular expressions
including equivalence classes, and the
.Xr strcoll 3
and
.Xr strxfrm 3
functions.
.It Ev LC_CTYPE
Intended to affect character encoding, character classification,
and case conversion.
For example, it is used by
.Xr mbtowc 3 ,
.Xr iswctype 3 ,
.Xr iswalnum 3 ,
.Xr towlower 3 ,
.Xr fgetwc 3 ,
.Xr fputwc 3 ,
.Xr printf 3 ,
and
.Xr scanf 3 .
.It Ev LC_MESSAGES
Intended to affect the output of informative and diagnostic messages
and the interpretation of interactive responses, in particular
regarding the language.
It is used by
.Xr catopen 3 .
.It Ev LC_MONETARY
Intended to affect monetary formatting.
.It Ev LC_NUMERIC
Intended to affect numeric, non-monetary formatting, for example
the radix character and thousands separators.
On other operating systems, it may for example affect
.Xr printf 3 ,
.Xr scanf 3 ,
and
.Xr strtod 3 .
.It Ev LC_TIME
Intended to affect date and time formats.
It may for example affect
.Xr strftime 3 .
.It Ev LANG
Fallback if any of the above is unset.
.It Ev NLSPATH
Used by
.Xr catopen 3
to locate message catalogs.
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa /usr/share/locale/UTF-8/LC_CTYPE
Character classification, case conversion, and character display
width database in
.Xr mklocale 1
binary output format used by
.Xr setlocale 3 .
.It Pa /usr/local/share/locale/
Localization data for
.Xr packages 7 ,
in particular
.Ev LC_MESSAGES
catalogs in GNU gettext format.
.It Pa /usr/local/share/nls/
Localization data for
.Xr packages 7 ,
in particular
.Ev LC_MESSAGES
catalogs in
.Xr catopen 3
format.
.It Pa /usr/src/share/locale/ctype/en_US.UTF-8.src
Character classification, case conversion, and character display
width database in
.Xr mklocale 1
input format.
.It Pa /usr/libdata/perl5/unicore/
Complete Unicode data used for generating the above database.
.It Pa /usr/src/gnu/usr.bin/perl/lib/unicore/UnicodeData.txt
The most important parts of Unicode data in a compact, more easily
human-readable format.
.El
.Sh EXIT STATUS
.Ex -std locale
.Sh SEE ALSO
.Xr setlocale 3
.Sh STANDARDS
With respect to locale support, most libraries and programs in the
.Ox
base system, including the
.Nm
utility, implement a subset of the
.St -p1003.1-2008
specification.
.Sh HISTORY
The
.Nm
utility was first standardized in the
.St -xpg4 .
.Pp
It was rewritten from scratch for
.Ox 5.4
during the 2013 Toronto hackathon.
.Sh AUTHORS
.An -nosplit
.An Stefan Sperling Aq Mt stsp@openbsd.org
with contributions from
.An Philip Guenther Aq Mt guenther@openbsd.org
and
.An Jeremie Courreges-Anglas Aq Mt jca@openbsd.org .
This manual page was written by
.An Ingo Schwarze Aq Mt schwarze@openbsd.org .
.Sh BUGS
The
.Nm
concept is inadequate for inter-process communication.
Two processes exchanging text, for example over a network, using
sockets, in shared memory, or even using plain text files always
need a protocol-specific way to negotiate the character encoding
used.
.Pp
The list of supported locales is perpetually incomplete.


ChangeLog
0.14 - ...
* OpenBSD headers now include their respective non-BSD headers where
appropriate
* Fixed uncommon pledge failures in `expr`, `cut`, `cmp` and `diff`

0.13 - Nov 6 2023
* Added OpenBSD `pledge()`
* Moved definitions to be compatible with libbsd
* Enabled pledge for most programs
* Fixed segfault in getopt

0.12 - Nov 1 2023
* Fixed numerous installation bugs
* Removed fts, cdefs and logwtmp from libcobalt
* Various small musl-to-glibc adaptions for libcobalt
* Added strlcpy, strlcat and fgetln to libcobalt
* GNUmakefile links with the local libcobalt directory
* Removed OpenBSD Make, build system replaced by GNU Make
* Removed m4, awk, getent, getconf and env
* Improved verbose `rm` text
* Move <sys/tree.h> header to <tree.h>
* Added pkg-config files
* libcobalt is now a drop-in replacement for libbsd and libmd
* Fixed `unicode` locale

0.11 - Sep 17 2023
* Removed orphaned binary
* Fixed mk file install location

0.10 - Sep 14 2023
* Removed `news`
* `unicode` and `freq` rewritten without Plan9 libraries
* Reworked `unicode` and `freq` man pages
* Removed all remaining plan9 code
* Removed dead source file from `ls` and fixed symlink reading
* Changed GNUmakefile references of libopenbsd to libcobalt
* Removed forced linking with libcobalt in mk files, which was causing
circular and/or unneccessary linkage
* `doas` is now correctly installed as setuid
* Removed support for static building from mk files
* Libraries built with mk files now always have a symlink without the soname
* Disabled mk files from checking for the crt
* Combined fortunes2 and the offensive variant into the main fortunes lists
* Added additional fortunes
* Added `wisdom` and `shsecret` programs
* `cp` no longer errors when `chmod` fails
* Build fixes

0.9 - Sep 9 2023
* Rename libopenbsd to libcobalt
* Set `cp -r` functionality identical to `cp -R`, allowing symlinks and
special files to be copied properly using more conventional syntax
* Disabled some useless or broken functionality in libcobalt
* Fixed memory corruption in siphash.c
* Removed srandom_deterministic
* Added base64, icdb, radixsort, md5, rmd160, sha1, sha2, sys/tree.h and
readpassphrase to libcobalt
* Cleaned up libcobalt makefile
* Removed dead symbols from openbsd.h
* Fixed build system
* Removed `yacc` in favor of Berkeley YACC
* Added `version` utility
* Ported doas (from OpenDoas) to cbtutils

0.8 - Sep 4 2023
* Added the following new utilities: cp mv arch awk banner basename
cat chroot cmp comm cut date dd diff dirname echo env expr false fold getconf
getopt head hostname indent jot kill ln logger logname m4 mkdir mktemp nl
printenv printf pwd readlink realpath make rev rm rmdir seq sleep split
tail timeout touch tr true tsort tty uname unifdef uniq unvis vis wc yacc yes
* Removed `c89`, `c99`, `bs`, `unstr`, `primes` and `unutf`
* Replaced Heirloom `banner` and NetBSD `rs` with OpenBSD counterparts
* Crunched crunchgen-ng into cbtutils
* Crunchgen mostly reimplemented as a compile-time skeleton
* Revamped build system for OpenBSD Make, with bootstrapping for GNU Make
* `strfile` from `fortune` is no longer installed
* Added -z flag to `ls` man page
* `units` now correctly flushes stdout
* Fully moved units.lib to /usr/share/misc/
* Removed dependency on libbsd
* Added man page for getent(1) and entropy(1)
* Removed useless lib9 header that conflicted with libopenbsd

0.7 - Aug 21 2023
* Add zxcvbn entropy program
* Add c89 and c99 compatibility programs for the C compiler
* Makefile cleanup
* Replaced unicode/ascii.1 with unicode/unicode.1

0.6 - Aug 5 2023
* Numerous makefile fixes
* Fixed cbtutils.conf
* Removed black background from ls entries
* Added getent

0.5 - Apr 26 2023
* Compilation fixes for mold linker
* Integrated support for crunchgen-ng
* Fixed libbio and lib9 linking
* Added a makefile install rule
* Fixed fortune/datfiles makefile

0.4 - Apr 19 2023
* Added proper man page for unstr
* Fixed broken references and inaccuracies in some man pages

0.3 - Apr 6 2023
* Removed nbperf
* Added support for any C compiler, instead of just gcc
* Makefile improvements for fortune
* Minor cleanup
* Removed astro
* Removed bcat executable from libbio
* Removed pbd
* Removed pom
* Fixed units not being built
* Removed worms
* Added manpages for libbio and strfile
* Added unstr

0.2 - Mar 24 2023
* Added bs
* Added grdc
* Added number
* Added pom
* Added rain
* Added tetris
* Added worms
* Added a license
* Various other improvements

0.1c - Mar 17 2023
* Fixed README.md

0.1b - Mar 17 2023
* Fixed README.md
* Added ChangeLog

0.1 - Mar 17 2023
* Added astro utility
* Added banner utility
* Added fortune utility
* Added freq utility
* Added lib9 library
* Added libbio library
* Added locale utility
* Added ls utility
* Added nbperf utility
* Added news utility
* Added pbd utility
* Added primes utility
* Added rs utility
* Added unicode utility
* Added units utility
* Added unutf utility


locale/locale.c
/*	$OpenBSD: locale.c,v 1.13 2020/04/17 10:45:04 schwarze Exp $	*/
/*
 * Copyright (c) 2013 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <err.h>
#include <langinfo.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

extern char		*__progname;

struct category_name {
	int category;
	const char *name;
} categories[] = {
	{ LC_COLLATE,	"LC_COLLATE" },
	{ LC_CTYPE,	"LC_CTYPE" },
	{ LC_MONETARY,	"LC_MONETARY" },
	{ LC_NUMERIC,	"LC_NUMERIC" },
	{ LC_TIME,	"LC_TIME" },
	{ LC_MESSAGES,	"LC_MESSAGES" },
	{ 0, 		NULL},
};

static void
put_assignment(const char *name, const char *value, int double_quoted)
{
	char c;

	fputs(name, stdout);
	putchar('=');
	if (double_quoted)
		putchar('"');
	if (value != NULL)
		while ((c = *value++) != '\0')
			switch (c) {
			case ' ': case '\t': case '\n': case '\'':
			case '(': case ')': case '<': case '>':
			case '&': case ';': case '|': case '~':
				if (!double_quoted)
			case '"': case '\\': case '$': case '`': 
					putchar('\\');
			default:
				putchar(c);
				break;
			}
	if (double_quoted)
		putchar('"');
	putchar('\n');
}

static void
show_current_locale(void)
{
	char *lang, *lc_all;
	int i;

	lang = getenv("LANG");
	lc_all = getenv("LC_ALL");

	put_assignment("LANG", lang, 0);
	for (i = 0; categories[i].name != NULL; i++) {
		if (lc_all == NULL && getenv(categories[i].name))
			put_assignment(categories[i].name,
			    getenv(categories[i].name), 0);
		else
			put_assignment(categories[i].name,
			    setlocale(categories[i].category, NULL), 1);
	}
	put_assignment("LC_ALL", lc_all, 0);
}

const char * const some_locales[] = {
	"C",
	"C.UTF-8",
	"POSIX",
	"POSIX.UTF-8",
	"Pig.UTF-8",
	"ar_SD.UTF-8",
	"ar_SY.UTF-8",
	"ca_ES.UTF-8",
	"cs_CZ.UTF-8",
	"da_DK.UTF-8",
	"de_AT.UTF-8",
	"de_CH.UTF-8",
	"de_DE.UTF-8",
	"el_GR.UTF-8",
	"en_AU.UTF-8",
	"en_CA.UTF-8",
	"en_GB.UTF-8",
	"en_US.UTF-8",
	"es_AR.UTF-8",
	"es_BO.UTF-8",
	"es_CH.UTF-8",
	"es_CO.UTF-8",
	"es_CR.UTF-8",
	"es_CU.UTF-8",
	"es_DO.UTF-8",
	"es_EC.UTF-8",
	"es_ES.UTF-8",
	"es_GQ.UTF-8",
	"es_GT.UTF-8",
	"es_HN.UTF-8",
	"es_MX.UTF-8",
	"es_NI.UTF-8",
	"es_PA.UTF-8",
	"es_PE.UTF-8",
	"es_PR.UTF-8",
	"es_PY.UTF-8",
	"es_SV.UTF-8",
	"es_US.UTF-8",
	"es_UY.UTF-8",
	"es_VE.UTF-8",
	"fa_IR.UTF-8",
	"fi_FI.UTF-8",
	"fr_BE.UTF-8",
	"fr_CA.UTF-8",
	"fr_CH.UTF-8",
	"fr_FR.UTF-8",
	"hu_HU.UTF-8",
	"hy_AM.UTF-8",
	"is_IS.UTF-8",
	"it_CH.UTF-8",
	"it_IT.UTF-8",
	"ja_JP.UTF-8",
	"ko_KR.UTF-8",
	"lt_LT.UTF-8",
	"nl_BE.UTF-8",
	"nl_NL.UTF-8",
	"no_NO.UTF-8",
	"pl_PL.UTF-8",
	"pt_PT.UTF-8",
	"ro_RO.UTF-8",
	"ru_RU.UTF-8",
	"sk_SK.UTF-8",
	"sl_SI.UTF-8",
	"sv_SE.UTF-8",
	"tr_TR.UTF-8",
	"uk_UA.UTF-8",
	"zh_CN.UTF-8",
	"zh_TW.UTF-8",
	NULL
};

static void
show_locales(void)
{
	int i = 0;

	while (some_locales[i])
		puts(some_locales[i++]);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [-a | -m | charmap]\n", __progname);
	exit(1);
}

int
main(int argc, char *argv[])
{
	int opt, aflag = 0, mflag = 0;

	setlocale(LC_ALL, "");

	if (argc == 1) {
		show_current_locale();
		return 0;
	}

	while ((opt = getopt(argc, argv, "am")) != -1) {
		switch (opt) {
		case 'a':
			aflag = 1;
			break;
		case 'm':
			mflag = 1;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (aflag + mflag + argc > 1)
		usage();
	else if (aflag)
		show_locales();
	else if (mflag)
		printf("UTF-8\n");
	else if (strcmp(*argv, "charmap") == 0)
		printf("%s\n", nl_langinfo(CODESET));
	else
		usage();

	return 0;
}
